## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [6]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [7]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [8]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [9]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [10]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()

top_3_categories = category_counts.head(3)
print(top_3_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [11]:
# For the category with the most entries, which subcategory had the most entries?
top_category = top_3_categories.index[0]

top_categories_df = df[df['category'] == top_category]

top_subcategory = top_categories_df['subcategory'].value_counts()

top_subcategory.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [12]:
# Which five clients had the most entries in the data?
client_counts = df['client_id'].value_counts()

client_counts.head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [13]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = client_counts.head(5).index.tolist()

top_5_client_ids

[33615, 66037, 46820, 38378, 24741]

In [14]:
# How many total units (the qty column) did the client with the most entries order order?
top_clients = df[df['client_id'] == top_5_client_ids[0]]

total_units = top_clients['qty'].sum()

total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [34]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']

print(df[['unit_price', 'qty', 'subtotal']].head(2))

   unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95


In [18]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']

df['shipping_price'] = df.apply(lambda row: row['total_weight'] * 7 if row['total_weight'] > 50 else row['total_weight'] * 10, axis = 1)

print(df[['unit_price','unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3))



   unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64


In [36]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
# Calculate the sales tax
df['sales_tax'] = (df['subtotal'] + df['shipping_price']) * 0.0925

# Calculate the total price
df['total_price'] = df['subtotal'] + df['shipping_price'] + df['sales_tax']

# Display the first few rows to verify the new columns
print(df[['subtotal', 'shipping_price', 'sales_tax', 'total_price']].head(3))


    subtotal  shipping_price     sales_tax    total_price
0  115164.00         5512.50  11162.576250  131839.076250
1     523.95          312.90     77.408625     914.258625
2     527.28          458.64     91.197600    1077.117600


In [19]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

# Calculate the cost of each line
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

# Display the first few rows to verify the new column
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   

   unit_cost  unit_weight  qty  line_number  total_weight  shipping_price  \
0     762.71         7.50  105            1        787.50         5512.50   
1      15.09         1.49   21            0         31.29          312.90   
2       7.86         1.68   39            6         65.52          458.64   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 21 columns]

In [39]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']

df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_weight  qty line_number   subtotal  \
0          17        2023  ...        7.50  105           1  115164.00   
1          20        2023  ...        1.49   21           0     523.95   
2           4        2023  ...        1.68   39           6     527.28   

   total_weight  shipping_price     sales_tax    total_price  line_cost  \
0        787.50         5512.50  11162.576250  131839.076250   85597.05   
1         31.29          312.90     77.408625     914.258625     629.79   
2         65.52          458.64     91.197600    1077.117600     765.18   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  

[3 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [40]:
# Check your work using the totals above

# Function to calculate total price for an order
def calculate_order_total(order_id):
    order_df = df[df['order_id'] == order_id]
    return order_df['total_price'].sum()

# Calculate totals for the given order IDs
order_ids = [2742071, 2173913, 6128929]
calculated_totals = {order_id: calculate_order_total(order_id) for order_id in order_ids}

# Print results
for order_id, total in calculated_totals.items():
    print(f"Order ID {order_id} calculated total: ${total:.2f}")

Order ID 2742071 calculated total: $152811.90
Order ID 2173913 calculated total: $162388.72
Order ID 6128929 calculated total: $923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [21]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

# Calculate total spending for each client
top_5_spending = {}
for client_id in top_5_client_ids:
    client_df = df[df['client_id'] == int(client_id)]
    total_spending = (client_df['unit_price'] * client_df['qty']).sum()
    top_5_spending[client_id] = total_spending


print("Spending for the top 5 clients by quantity:")
for client_id, spending in top_5_spending.items():
    print(f"Client ID {client_id}: ${spending:.2f}")


Spending for the top 5 clients by quantity:
Client ID 33615: $5839032.11
Client ID 66037: $7995708.38
Client ID 46820: $7317356.03
Client ID 38378: $8384321.15
Client ID 24741: $70176885.25


In [84]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

# Calculate the required totals for each client

summary_data = []
for client_id in top_5_client_ids:
    client_df = df[df['client_id'] == int(client_id)]
    
    total_units = client_df['qty'].sum()
    total_cost = client_df['unit_cost'] * client_df['qty']
    total_shipping = client_df['shipping_price'].sum()
    total_revenue = (client_df['unit_price'] * client_df['qty']).sum()
    total_profit = (total_revenue - total_cost - total_shipping).sum()
    
    summary_data.append({
        'client_id': client_id,
        'total_units': total_units,
        'total_shipping': total_shipping,
        'total_revenue': total_revenue,
        'total_profit': total_profit
    })

# Create the summary DataFrame
summary_df = pd.DataFrame(summary_data)

# Set the client_id as the index
summary_df.set_index('client_id', inplace=True)

# Display the summary DataFrame
print(summary_df)


           total_units  total_shipping  total_revenue  total_profit
client_id                                                          
33615            64313      1828984.89     5839032.11  8.778641e+08
66037            43018      1395151.85     7995708.38  1.387108e+09
46820            75768      1601448.84     7317356.03  1.189219e+09
38378            73667      3429455.40     8384321.15  1.019452e+09
24741           239862      5126448.37    70176885.25  1.342488e+10


In [85]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['total_shipping', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    return f"${value / 1000000:.2f}M"

# Apply the currency_format_millions function to only the money columns. 
for col in money_columns:
    summary_df[col] = summary_df[col].apply(currency_format_millions)


# Rename the columns to reflect the change in the money format. 
summary_df = summary_df.rename(columns={'client_id': 'Client ID', 
                                        'total_units': 'Total Units', 
                                        'total_shipping': 'Shipping (Millions)', 
                                        'total_revenue': 'Total Revenue (Millions)', 
                                        'total_profit': 'Total Profit (Millions)'
                                        })

# Set the Client ID as the index
#summary_df.set_index('Client ID', inplace=True)

# Display the formatted summary DataFrame
print(summary_df)



           Total Units Shipping (Millions) Total Revenue (Millions)  \
client_id                                                             
33615            64313              $1.83M                   $5.84M   
66037            43018              $1.40M                   $8.00M   
46820            75768              $1.60M                   $7.32M   
38378            73667              $3.43M                   $8.38M   
24741           239862              $5.13M                  $70.18M   

          Total Profit (Millions)  
client_id                          
33615                    $877.86M  
66037                   $1387.11M  
46820                   $1189.22M  
38378                   $1019.45M  
24741                  $13424.88M  


In [86]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df = summary_df.sort_values(by='Total Profit (Millions)', ascending=False)
sorted_df

Total Units Shipping (Millions) Total Revenue (Millions)  \
client_id                                                             
33615            64313              $1.83M                   $5.84M   
66037            43018              $1.40M                   $8.00M   
24741           239862              $5.13M                  $70.18M   
46820            75768              $1.60M                   $7.32M   
38378            73667              $3.43M                   $8.38M   

          Total Profit (Millions)  
client_id                          
33615                    $877.86M  
66037                   $1387.11M  
24741                  $13424.88M  
46820                   $1189.22M  
38378                   $1019.45M